Repeated QuotaExceededError

### What version of Hls.js are you using?

v1.5.11

### What browser (including version) are you using?

Chrome 126.0.6478.127

### What OS (including version) are you using?

Windows 11

### Test stream

https://hlsjs.video-dev.org/demo/?src=https%3A%2F%2Fsample-videos-zyrkp2nj.s3-eu-west-1.amazonaws.com%2Fbig-buck-bunny%2Fhls%2Fmaster.m3u8&demoConfig=eyJlbmFibGVTdHJlYW1pbmciOnRydWUsImF1dG9SZWNvdmVyRXJyb3IiOnRydWUsInN0b3BPblN0YWxsIjpmYWxzZSwiZHVtcGZNUDQiOmZhbHNlLCJsZXZlbENhcHBpbmciOi0xLCJsaW1pdE1ldHJpY3MiOi0xfQ==

### Configuration

```JavaScript
{
  "debug": true,
  "startLevel": 0,
  "capLevelToPlayerSize": true
}
```


### Additional player setup steps

- Set `HTML video element width` to `240px`
- Load stream

Alternatively, without the config:
- Load stream
- Pause
- Manually select lowest rendition

### Checklist

- [X] The issue observed is not already reported by searching on Github under https://github.com/video-dev/hls.js/issues
- [X] The issue occurs in the stable client (latest release) on https://hlsjs.video-dev.org/demo and not just on my page
- [X] The issue occurs in the latest client (main branch) on https://hlsjs-dev.video-dev.org/demo and not just on my page
- [X] The stream has correct Access-Control-Allow-Origin headers (CORS)
- [X] There are no network errors such as 404s in the browser console when trying to play the stream

### Steps to reproduce

1. Load the provided stream
2. Manually select lowest resolution
3. Pause the video


### Expected behaviour

Buffer size should be restricted on `QuotaExceededError`.

### What actually happened?

Due to playing very low bitrate video, the audio buffer overfills before reaching the limits of the video buffer. This leads to repeated `QuotaExceededErrors: Failed to execute 'appendBuffer' on 'SourceBuffer': The SourceBuffer is full, and cannot free space to append additional buffers.` being triggered, and the last audio segment being repeatedly downloaded.

### Console output

```shell
Clipped as it's infinitely repeating.

[log] > [stream-controller]: IDLE->FRAG_LOADING
transmuxer-interface.ts:230 [log] > [transmuxer-interface, audio]: Starting new transmux session for sn: 91 p: -1 level: 0 id: 1
        discontinuity: false
        trackSwitch: false
        contiguous: false
        accurateTimeOffset: false
        timeOffset: 546.0193673470475
        initSegmentChange: false
base-stream-controller.ts:389 [log] > [audio-stream-controller]: Loaded fragment 91 of level 0
transmuxer-interface.ts:394 [log] > [mp4-remuxer]: reset next timestamp
base-stream-controller.ts:1793 [log] > [audio-stream-controller]: FRAG_LOADING->PARSING
buffer-operation-queue.ts:64 [warn] > [buffer-operation-queue]: Exception executing "audio" SourceBuffer operation: QuotaExceededError: Failed to execute 'appendBuffer' on 'SourceBuffer': The SourceBuffer is full, and cannot free space to append additional buffers.
executeNext @ buffer-operation-queue.ts:64
append @ buffer-operation-queue.ts:29
onBufferAppending @ buffer-controller.ts:548
emit @ index.js:203
emit @ hls.ts:310
trigger @ hls.ts:318
bufferFragmentData @ base-stream-controller.ts:943
_handleTransmuxComplete @ audio-stream-controller.ts:850
handleTransmuxComplete @ transmuxer-interface.ts:422
onWorkerMessage @ transmuxer-interface.ts:382
TransmuxerInterface.onwmsg @ transmuxer-interface.ts:91
Show 1 more frame
Show less
base-stream-controller.ts:1651 [log] > [audio-stream-controller]: Reset loading state
base-stream-controller.ts:1793 [log] > [audio-stream-controller]: PARSING->IDLE
main.js:734 Error event: {type: 'mediaError', parent: 'audio', details: 'bufferFullError', sourceBufferName: 'audio', frag: Fragment, …}
(anonymous) @ main.js:734
emit @ index.js:203
emit @ hls.ts:310
trigger @ hls.ts:318
onError @ buffer-controller.ts:545
executeNext @ buffer-operation-queue.ts:67
append @ buffer-operation-queue.ts:29
onBufferAppending @ buffer-controller.ts:548
emit @ index.js:203
emit @ hls.ts:310
trigger @ hls.ts:318
bufferFragmentData @ base-stream-controller.ts:943
_handleTransmuxComplete @ audio-stream-controller.ts:850
handleTransmuxComplete @ transmuxer-interface.ts:422
onWorkerMessage @ transmuxer-interface.ts:382
TransmuxerInterface.onwmsg @ transmuxer-interface.ts:91
Show 2 more frames
Show less
transmuxer-interface.ts:394 [log] > [transmuxer.ts]: Flushed fragment 91 of level 0
base-stream-controller.ts:758 [log] > [audio-stream-controller]: Loading fragment 91 cc: 0 of [0-99] track: 0, target: 546.019
base-stream-controller.ts:1793 [log] > [audio-stream-controller]: IDLE->FRAG_LOADING
base-stream-controller.ts:389 [log] > [stream-controller]: Loaded fragment 98 of level 0
base-stream-controller.ts:1793 [log] > [stream-controller]: FRAG_LOADING->PARSING
transmuxer-interface.ts:394 [log] > [transmuxer.ts]: Flushed fragment 98 of level 0
base-stream-controller.ts:1793 [log] > [stream-controller]: PARSING->PARSED
base-stream-controller.ts:570 [log] > [stream-controller]: Buffered main sn: 98 of level 0 (frag:[588.000-594.000] > buffer:[0.000-594.000])
base-stream-controller.ts:1793 [log] > [stream-controller]: PARSED->IDLE
base-stream-controller.ts:758 [log] > [stream-controller]: Loading fragment 99 cc: 0 of [0-99] level: 0, target: 594
base-stream-controller.ts:1793 [log] > [stream-controller]: IDLE->FRAG_LOADING
transmuxer-interface.ts:230 [log] > [transmuxer-interface, audio]: Starting new transmux session for sn: 91 p: -1 level: 0 id: 1
        discontinuity: false
        trackSwitch: false
        contiguous: false
        accurateTimeOffset: false
        timeOffset: 546.0193673470475
        initSegmentChange: false
base-stream-controller.ts:389 [log] > [audio-stream-controller]: Loaded fragment 91 of level 0
transmuxer-interface.ts:394 [log] > [mp4-remuxer]: reset next timestamp
base-stream-controller.ts:1793 [log] > [audio-stream-controller]: FRAG_LOADING->PARSING
buffer-operation-queue.ts:64 [warn] > [buffer-operation-queue]: Exception executing "audio" SourceBuffer operation: QuotaExceededError: Failed to execute 'appendBuffer' on 'SourceBuffer': The SourceBuffer is full, and cannot free space to append additional buffers.
executeNext @ buffer-operation-queue.ts:64
append @ buffer-operation-queue.ts:29
onBufferAppending @ buffer-controller.ts:548
emit @ index.js:203
emit @ hls.ts:310
trigger @ hls.ts:318
bufferFragmentData @ base-stream-controller.ts:943
_handleTransmuxComplete @ audio-stream-controller.ts:850
handleTransmuxComplete @ transmuxer-interface.ts:422
onWorkerMessage @ transmuxer-interface.ts:382
TransmuxerInterface.onwmsg @ transmuxer-interface.ts:91
Show 1 more frame
Show less
base-stream-controller.ts:1651 [log] > [audio-stream-controller]: Reset loading state
base-stream-controller.ts:1793 [log] > [audio-stream-controller]: PARSING->IDLE
main.js:734 Error event: {type: 'mediaError', parent: 'audio', details: 'bufferFullError', sourceBufferName: 'audio', frag: Fragment, …}
(anonymous) @ main.js:734
emit @ index.js:203
emit @ hls.ts:310
trigger @ hls.ts:318
onError @ buffer-controller.ts:545
executeNext @ buffer-operation-queue.ts:67
append @ buffer-operation-queue.ts:29
onBufferAppending @ buffer-controller.ts:548
emit @ index.js:203
emit @ hls.ts:310
trigger @ hls.ts:318
bufferFragmentData @ base-stream-controller.ts:943
_handleTransmuxComplete @ audio-stream-controller.ts:850
handleTransmuxComplete @ transmuxer-interface.ts:422
onWorkerMessage @ transmuxer-interface.ts:382
TransmuxerInterface.onwmsg @ transmuxer-interface.ts:91
Show 2 more frames
Show less
transmuxer-interface.ts:394 [log] > [transmuxer.ts]: Flushed fragment 91 of level 0
base-stream-controller.ts:758 [log] > [audio-stream-controller]: Loading fragment 91 cc: 0 of [0-99] track: 0, target: 546.019
base-stream-controller.ts:1793 [log] > [audio-stream-controller]: IDLE->FRAG_LOADING
base-stream-controller.ts:389 [log] > [stream-controller]: Loaded fragment 99 of level 0
base-stream-controller.ts:1793 [log] > [stream-controller]: FRAG_LOADING->PARSING
transmuxer-interface.ts:394 [log] > [transmuxer.ts]: Flushed fragment 99 of level 0
base-stream-controller.ts:1793 [log] > [stream-controller]: PARSING->PARSED
base-stream-controller.ts:570 [log] > [stream-controller]: Buffered main sn: 99 of level 0 (frag:[594.000-596.440] > buffer:[0.000-596.440])
base-stream-controller.ts:1793 [log] > [stream-controller]: PARSED->IDLE
buffer-controller.ts:638 [log] > [buffer-controller] video sourceBuffer now EOS
base-stream-controller.ts:1793 [log] > [stream-controller]: IDLE->ENDED
transmuxer-interface.ts:230 [log] > [transmuxer-interface, audio]: Starting new transmux session for sn: 91 p: -1 level: 0 id: 1
        discontinuity: false
        trackSwitch: false
        contiguous: false
        accurateTimeOffset: false
        timeOffset: 546.0193673470475
        initSegmentChange: false
base-stream-controller.ts:389 [log] > [audio-stream-controller]: Loaded fragment 91 of level 0
transmuxer-interface.ts:394 [log] > [mp4-remuxer]: reset next timestamp
base-stream-controller.ts:1793 [log] > [audio-stream-controller]: FRAG_LOADING->PARSING
buffer-operation-queue.ts:64 [warn] > [buffer-operation-queue]: Exception executing "audio" SourceBuffer operation: QuotaExceededError: Failed to execute 'appendBuffer' on 'SourceBuffer': The SourceBuffer is full, and cannot free space to append additional buffers.
executeNext @ buffer-operation-queue.ts:64
append @ buffer-operation-queue.ts:29
onBufferAppending @ buffer-controller.ts:548
emit @ index.js:203
emit @ hls.ts:310
trigger @ hls.ts:318
bufferFragmentData @ base-stream-controller.ts:943
_handleTransmuxComplete @ audio-stream-controller.ts:850
handleTransmuxComplete @ transmuxer-interface.ts:422
onWorkerMessage @ transmuxer-interface.ts:382
TransmuxerInterface.onwmsg @ transmuxer-interface.ts:91
Show 1 more frame
Show less
base-stream-controller.ts:1651 [log] > [audio-stream-controller]: Reset loading state
base-stream-controller.ts:1793 [log] > [audio-stream-controller]: PARSING->IDLE
main.js:734 Error event: {type: 'mediaError', parent: 'audio', details: 'bufferFullError', sourceBufferName: 'audio', frag: Fragment, …}
(anonymous) @ main.js:734
emit @ index.js:203
emit @ hls.ts:310
trigger @ hls.ts:318
onError @ buffer-controller.ts:545
executeNext @ buffer-operation-queue.ts:67
append @ buffer-operation-queue.ts:29
onBufferAppending @ buffer-controller.ts:548
emit @ index.js:203
emit @ hls.ts:310
trigger @ hls.ts:318
bufferFragmentData @ base-stream-controller.ts:943
_handleTransmuxComplete @ audio-stream-controller.ts:850
handleTransmuxComplete @ transmuxer-interface.ts:422
onWorkerMessage @ transmuxer-interface.ts:382
TransmuxerInterface.onwmsg @ transmuxer-interface.ts:91
Show 2 more frames
Show less
transmuxer-interface.ts:394 [log] > [transmuxer.ts]: Flushed fragment 91 of level 0
base-stream-controller.ts:758 [log] > [audio-stream-controller]: Loading fragment 91 cc: 0 of [0-99] track: 0, target: 546.019
base-stream-controller.ts:1793 [log] > [audio-stream-controller]: IDLE->FRAG_LOADING
transmuxer-interface.ts:230 [log] > [transmuxer-interface, audio]: Starting new transmux session for sn: 91 p: -1 level: 0 id: 1
        discontinuity: false
        trackSwitch: false
        contiguous: false
        accurateTimeOffset: false
        timeOffset: 546.0193673470475
        initSegmentChange: false
```
```


### Chrome media internals output

_No response_

--- a/api-extractor/report/hls.js.api.md
+++ b/api-extractor/report/hls.js.api.md
@@ -459,7 +459,7 @@ export class BaseStreamController extends TaskLoop implements NetworkComponentAP
     // (undocumented)
     protected reduceLengthAndFlushBuffer(data: ErrorData): boolean;
     // (undocumented)
-    protected reduceMaxBufferLength(threshold: number): boolean;
+    protected reduceMaxBufferLength(threshold: number, fragDuration: number): boolean;
     // (undocumented)
     protected removeUnbufferedFrags(start?: number): void;
     // (undocumented)

--- a/src/controller/base-stream-controller.ts
+++ b/src/controller/base-stream-controller.ts
@@ -425,7 +425,7 @@ export default class BaseStreamController
         : 0;
       if (
         backtracked === 1 ||
-        this.reduceMaxBufferLength(minForwardBufferLength)
+        this.reduceMaxBufferLength(minForwardBufferLength, frag.duration)
       ) {
         fragmentTracker.removeFragment(frag);
       }
@@ -1026,10 +1026,16 @@ export default class BaseStreamController
     return Math.min(maxBufLen, config.maxMaxBufferLength);
   }
 
-  protected reduceMaxBufferLength(threshold: number) {
+  protected reduceMaxBufferLength(threshold: number, fragDuration: number) {
     const config = this.config;
-    const minLength = threshold || config.maxBufferLength;
-    const reducedLength = config.maxMaxBufferLength / 2;
+    const minLength = Math.max(
+      Math.min(threshold, config.maxBufferLength),
+      fragDuration,
+    );
+    const reducedLength = Math.max(
+      threshold - fragDuration * 3,
+      config.maxMaxBufferLength / 2,
+    );
     if (reducedLength >= minLength) {
       // reduce max buffer length as it might be too high. we do this to avoid loop flushing ...
       config.maxMaxBufferLength = reducedLength;
@@ -1581,6 +1587,7 @@ export default class BaseStreamController
   protected reduceLengthAndFlushBuffer(data: ErrorData): boolean {
     // if in appending state
     if (this.state === State.PARSING || this.state === State.PARSED) {
+      const frag = data.frag;
       const playlistType = data.parent as PlaylistLevelType;
       const bufferedInfo = this.getFwdBufferInfo(
         this.mediaBuffer,
@@ -1590,7 +1597,7 @@ export default class BaseStreamController
       // reduce max buf len if current position is buffered
       const buffered = bufferedInfo && bufferedInfo.len > 0.5;
       if (buffered) {
-        this.reduceMaxBufferLength(bufferedInfo.len);
+        this.reduceMaxBufferLength(bufferedInfo.len, frag?.duration || 10);
       }
       const flushBuffer = !buffered;
       if (flushBuffer) {
@@ -1601,9 +1608,9 @@ export default class BaseStreamController
           `Buffer full error while media.currentTime is not buffered, flush ${playlistType} buffer`,
         );
       }
-      if (data.frag) {
-        this.fragmentTracker.removeFragment(data.frag);
-        this.nextLoadPosition = data.frag.start;
+      if (frag) {
+        this.fragmentTracker.removeFragment(frag);
+        this.nextLoadPosition = frag.start;
       }
       this.resetLoadingState();
       return flushBuffer;

